src/mds/MDSRank.h

```diff
@@ -409,7 +409,7 @@ class MDSRank {
     // The last different state I held before current
     MDSMap::DaemonState last_state = MDSMap::STATE_BOOT;
     // The state assigned to me by the MDSMap
-    MDSMap::DaemonState state = MDSMap::STATE_BOOT;
+    MDSMap::DaemonState state = MDSMap::STATE_STANDBY;
```

Comment:
\
```suggestion
    MDSMap::DaemonState state = MDSMap::STATE_NULL;
```

would be more appropriate I think.

Reply:
if it is not `MDSMap::STATE_STANDBY`, the [added assert](https://github.com/ceph/ceph/pull/44501/files#diff-f91675b3e095c571fa2e401547ac27a8f0b1575788f0df3321a8f20a5949f90dR2242) would fail, or the [`MDSMap::state_transition_valid(oldstate, state)`](https://github.com/ceph/ceph/pull/44501/files#diff-f91675b3e095c571fa2e401547ac27a8f0b1575788f0df3321a8f20a5949f90dR2261) would return false.

Reply:
I think @batrick is right - This causes an failed assert::

```
    -5> 2022-02-05T21:48:44.562+0000 7fc710419700  4 mds.0.0 apply_blocklist: killed 0 blocklisted sessions (0 blocklist entries, 0)
    -4> 2022-02-05T21:48:44.562+0000 7fc710419700 10 mds.c handle_mds_map: handling map as rank 0
    -3> 2022-02-05T21:48:44.562+0000 7fc710419700 -1 /build/ceph-17.0.0-10531-g09fc2e9e/src/mds/MDSRank.cc: In function 'void MDSRankDispatcher::handle_mds_map(ceph::cref_t<MMDSMap>&, const MDSMap&)' thread 7fc710419700 time 2022-02-05T21:48:44.566988+0000
/build/ceph-17.0.0-10531-g09fc2e9e/src/mds/MDSRank.cc: 2242: FAILED ceph_assert(state == oldmap.get_state_gid(mds_gid))

 ceph version 17.0.0-10531-g09fc2e9e (09fc2e9e470435f1e6dbf462a51853391f4a4cc0) quincy (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x14f) [0x7fc71593c7cc]
 2: /usr/lib/ceph/libceph-common.so.2(+0x27b9de) [0x7fc71593c9de]
 3: (MDSRankDispatcher::handle_mds_map(boost::intrusive_ptr<MMDSMap const> const&, MDSMap const&)+0x2df7) [0x7fc716684dc7]
 4: (MDSDaemon::handle_mds_map(boost::intrusive_ptr<MMDSMap const> const&)+0x107b) [0x7fc71665b67b]
 5: (MDSDaemon::handle_core_message(boost::intrusive_ptr<Message const> const&)+0x3df) [0x7fc71665d1ef]
 6: (MDSDaemon::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x186) [0x7fc71665db86]
 7: (Messenger::ms_deliver_dispatch(boost::intrusive_ptr<Message> const&)+0x460) [0x7fc715bd48d0]
 8: (DispatchQueue::entry()+0x58f) [0x7fc715bd216f]
 9: (DispatchQueue::DispatchThread::entry()+0x11) [0x7fc715ca0441]
 10: /lib/x86_64-linux-gnu/libpthread.so.0(+0x9609) [0x7fc71566c609]
 11: clone()
```

Happens when an MDSRank gets initialized for the very first time - https://pulpito.ceph.com/vshankar-2022-02-05_17:27:49-fs-wip-vshankar-testing-20220201-113815-testing-default-smithi/6663006/

(You may want to vet this with your other changes in the PR).

Reply:
This failing assert is added in this PR.

I assumed MDSRank would only be initialized when the MDSDaemon is in STANDBY state. This should be true according to the [state diagram](https://docs.ceph.com/en/pacific/cephfs/mds-states/#lines). After the changes in this PR, MDS should reject any state change that is not in the diagram.

I will take a look. Do you have any idea why the MDS would initialize MDSRank while not in STANDBY state? Or how can I reproduce this locally?

Or maybe I should just relax the assert, and allow MDSRank to be initialized in any state.

Reply:
OK, I've managed to reproduce this by setting `paxos_propose_interval` to 5.0 then restarting MDS. Will fix this tomorrow.

Reply:
thx @huww98